src/D2L.CodeStyle.Analyzers/Immutability/ReadOnlyParameterAnalyzer.cs

```diff
@@ -0,0 +1,117 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class ReadOnlyParameterAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.ReadOnlyParameterIsnt
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private static void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+
+			Compilation compilation = context.Compilation;
+
+			INamedTypeSymbol readOnlyAttribute = compilation.GetTypeByMetadataName( "D2L.CodeStyle.Annotations.ReadOnlyAttribute" );
+
+			if( readOnlyAttribute == null ) {
+				return;
+			}
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeParameter(
+					ctx: ctx,
+					readOnlyAttribute: readOnlyAttribute,
+					syntax: ctx.Node as ParameterSyntax
+				),
+				SyntaxKind.Parameter
+			);
+		}
+
+		private static void AnalyzeParameter(
+			SyntaxNodeAnalysisContext ctx,
+			INamedTypeSymbol readOnlyAttribute,
+			ParameterSyntax syntax
+		) {
+			SemanticModel model = ctx.SemanticModel;
+
+			IParameterSymbol parameter = model.GetDeclaredSymbol( syntax, ctx.CancellationToken );
+
+			if( !IsMarkedReadOnly( readOnlyAttribute, parameter ) ) {
+				return;
+			}
+
+			if( parameter.RefKind != RefKind.None ) {
+				/**
+				 * public void Foo( [ReadOnly] in int foo )
+				 * public void Foo( [ReadOnly] ref int foo )
+				 * public void Foo( [ReadOnly] out int foo )
+				 */
+				ctx.ReportDiagnostic( Diagnostic.Create(
+					Diagnostics.ReadOnlyParameterIsnt,
+					syntax.GetLocation(),
+					"is an in/ref/out parameter"
+				) );
+			}
+
+			IMethodSymbol method = parameter.ContainingSymbol as IMethodSymbol;
+			BlockSyntax methodBody = ( method.DeclaringSyntaxReferences[0].GetSyntax( ctx.CancellationToken ) as MethodDeclarationSyntax ).Body;
+
+			DataFlowAnalysis dataflow = model.AnalyzeDataFlow( methodBody );
+			if( dataflow.WrittenInside.Contains( parameter ) ) {
+				/**
+				 * public void Foo( [ReadOnly] int foo ) {

```

Comment:
So why are we worried about this? Attributes on parameters are kind of "for the caller", and the caller shouldn't need to care about this... there is no way they can observe this happening and its equivalent to assigning foo to bar and then manipulating bar. I'm not sure why we want to prevent harmless code like this?

Reply:
Harmless, and the alternative is just more verbose

Reply:
Maybe the hang-up for me is just that we're not filtering out value types? (`delegate`/`Func`/etc. are reference types)
But in the reply about `in` you mentioned that it not working for value types was bad. I guess I'm not sure why.